src/listener/mod.rs

```diff
@@ -26,18 +30,46 @@ pub(crate) use tcp_listener::TcpListener;
 #[cfg(all(unix, feature = "h1-server"))]
 pub(crate) use unix_listener::UnixListener;
 
-/// The Listener trait represents an implementation of http transport
-/// for a tide application. In order to provide a Listener to tide,
-/// you will also need to implement at least one [`ToListener`](crate::listener::ToListener) that
+/// The Listener trait represents an implementation of http transport for a tide
+/// application. In order to provide a Listener to tide, you will also need to
+/// implement at least one [`ToListener`](crate::listener::ToListener) that
 /// outputs your Listener type.
-#[async_trait::async_trait]
-pub trait Listener<State: 'static>:
-    std::fmt::Debug + std::fmt::Display + Send + Sync + 'static
+#[async_trait]
+pub trait Listener<State>: Debug + Display + Send + Sync + 'static
+where
+    State: Send + Sync + 'static,
+{
+    /// Bind the listener. This starts the listening process by opening the
+    /// necessary network ports, but not yet accepting incoming connections. This
+    /// method must be called before `accept`.
+    async fn bind(&mut self, app: Server<State>) -> io::Result<()>;
+
+    /// Start accepting incoming connections. This method must be called only
+    /// after `bind` has succeeded.
+    async fn accept(&mut self) -> io::Result<()>;
```

Comment:
What if we put this function on something that `bind` returns? Or would that need to consume the app defeating the point?

Reply:
Actually I think `Server::bind()` already does this.